src/go/k8s/api/redpanda/v1alpha2/user_types.go

```diff
@@ -0,0 +1,196 @@
+func (u *User) RedpandaName() string {
+	if u.Spec.UsernameOverride != nil {
+		return *u.Spec.UsernameOverride
+	}
+	return fmt.Sprintf("%s/%s", u.Namespace, u.Name)
+}
+
+// ACLName constructs the name of a User for defining ACLs.
+func (u *User) ACLName() string {
+	return "User:" + u.RedpandaName()
+}
+
+func (u *User) GetKafkaAPISpec() *KafkaAPISpec {
+	return u.Spec.KafkaAPISpec
+}
+
+func (u *User) GetAdminAPISpec() *AdminAPISpec {
+	return u.Spec.AdminAPISpec
+}
+
+func (u *User) GetClusterRef() *ClusterRef {
+	return u.Spec.ClusterRef
+}
+
+// UserSpec defines the configuration of a Redpanda user.
+// +kubebuilder:validation:XValidation:message="either clusterRef or kafkaApiSpec and adminApiSpec must be set",rule="has(self.clusterRef) || (has(self.kafkaApiSpec) && has(self.adminApiSpec))"
+type UserSpec struct {
+	// ClusterRef is a reference to the cluster where the user should be created.
+	// It is used in constructing the client created to configure a cluster.
+	// This takes precedence over KafkaAPISpec and AdminAPISpec.
+	ClusterRef *ClusterRef `json:"clusterRef,omitempty"`
+	// KafkaAPISpec is the configuration information for communicating with the Kafka
+	// API of a Redpanda cluster where the User should be created.
+	KafkaAPISpec *KafkaAPISpec `json:"kafkaApiSpec,omitempty"`
+	// AdminAPISpec is the configuration information for communicating with the Admin
+	// API of a Redpanda cluster where the User should be created.
+	AdminAPISpec *AdminAPISpec `json:"adminApiSpec,omitempty"`
+	// Authentication defines the authentication information for a user. If no
+	// Authentication credentials are specified, then no user will be created.
+	// This is useful when wanting to manage ACLs for an already-existing user.
+	Authentication *UserAuthenticationSpec `json:"authentication,omitempty"`
+	// Authorization rules defined for this user.
+	Authorization *UserAuthorizationSpec `json:"authorization,omitempty"`
+	// Template to specify how user secrets are generated.
+	Template *UserTemplateSpec `json:"template,omitempty"`
+	// UsernameOverride allows for specifying a particular username for the
+	// given user rather than the calculated "<NAMESPACE>/<NAME>".
+	UsernameOverride *string `json:"usernameOverride,omitempty"`
+}
+
+// UserTemplateSpec defines the template metadata for a user
+type UserTemplateSpec struct {
+	// Specifies how the Secret with a user password is generated.
+	Secret *ResourceTemplate `json:"secret,omitempty"`
+}
+
+// UserAuthenticationSpec defines the authentication mechanism enabled for this Redpanda user.
+type UserAuthenticationSpec struct {
+	// +kubebuilder:validation:Enum=scram-sha-256;scram-sha-512
+	// +default:value=scram-sha-512
+	Type *string `json:"type,omitempty"`
+	// Password specifies where a password is read from.
+	Password Password `json:"password"`
+}
+
+type Password struct {
+	// +required
+	ValueFrom *PasswordSource `json:"valueFrom"`
```

Comment:
Yeah, so just going off the shape of Strimzi stuff, I think that this was designed after stuff like [`EnvVar`](https://github.com/kubernetes/kubernetes/blob/9edcffcde5595e8a5b1a35f88c421764e575afce/pkg/apis/core/types.go#L2132-L2153) where you have:

```
Value string
ValueFrom *EnvVarSource
```

and `EnvVarSource` has:

```
FieldRef *ObjectFieldSelector
ResourceFieldRef *ResourceFieldSelector
ConfigMapKeyRef *ConfigMapKeySelector
SecretKeyRef *SecretKeySelector
```

Surprisingly, they didn't include anything more than `ValueFrom` and `SecretKeyRef` though according to their docs, which is just what I followed. I can change it though to include at minimum `Value string` though, but the pointer I think is because it's envisioning alternative sources/values.

Reply:
Added a `Value` field.